File: android/src/main/java/com/vinay/network_speed_plus/NetworkSpeedPlusPlugin.java
package com.vinay.network_speed_plus;

import android.content.Context;
import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.EventChannel;

public class NetworkSpeedPlusPlugin implements FlutterPlugin, MethodCallHandler {
    private static final String METHOD_CHANNEL = "com.vinay.network_speed_plus/methods";
    private static final String EVENT_CHANNEL = "com.vinay.network_speed_plus/events";

    private MethodChannel methodChannel;
    private EventChannel eventChannel;
    private NetworkSpeedHandler speedHandler;
    private Context context;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        context = binding.getApplicationContext();
        
        methodChannel = new MethodChannel(binding.getBinaryMessenger(), METHOD_CHANNEL);
        methodChannel.setMethodCallHandler(this);

        eventChannel = new EventChannel(binding.getBinaryMessenger(), EVENT_CHANNEL);
        speedHandler = new NetworkSpeedHandler(context);
        eventChannel.setStreamHandler(speedHandler);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        switch (call.method) {
            case "startMonitoring":
                Boolean monitorTotalTraffic = call.argument("monitorTotalTraffic");
                Integer updateIntervalSeconds = call.argument("updateIntervalSeconds");
                if (monitorTotalTraffic != null && updateIntervalSeconds != null) {
                    speedHandler.startMonitoring(monitorTotalTraffic, updateIntervalSeconds);
                    result.success(true);
                } else {
                    result.error("INVALID_ARGS", "Missing arguments", null);
                }
                break;

            case "stopMonitoring":
                speedHandler.stopMonitoring();
                result.success(true);
                break;

            case "isMonitoring":
                result.success(speedHandler.isMonitoring());
                break;

            case "getMonitoringMode":
                result.success(speedHandler.getMonitoringMode());
                break;

            case "getUpdateInterval":
                result.success(speedHandler.getUpdateInterval());
                break;

            case "setUpdateInterval":
                Integer seconds = call.arguments();
                if (seconds != null) {
                    speedHandler.setUpdateInterval(seconds);
                    result.success(true);
                } else {
                    result.error("INVALID_ARGS", "Missing seconds argument", null);
                }
                break;

            case "dispose":
                speedHandler.dispose();
                result.success(true);
                break;

            default:
                result.notImplemented();
                break;
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        if (methodChannel != null) {
            methodChannel.setMethodCallHandler(null);
            methodChannel = null;
        }
        if (eventChannel != null) {
            eventChannel.setStreamHandler(null);
            eventChannel = null;
        }
        if (speedHandler != null) {
            speedHandler.dispose();
            speedHandler = null;
        }
        context = null;
    }
}

// import io.flutter.embedding.engine.plugins.FlutterPlugin;
// import io.flutter.plugin.common.MethodCall;
// import io.flutter.plugin.common.MethodChannel;
// import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
// import io.flutter.plugin.common.MethodChannel.Result;
// import io.flutter.plugin.common.EventChannel;
// import io.flutter.plugin.common.PluginRegistry.Registrar;

// public class NetworkSpeedPlusPlugin implements FlutterPlugin, MethodCallHandler {
//   private static final String METHOD_CHANNEL = "com.vinay.network_speed_plus/methods";
//   private static final String EVENT_CHANNEL = "com.vinay.network_speed_plus/events";

//   private MethodChannel methodChannel;
//   private EventChannel eventChannel;
//   private NetworkSpeedHandler speedHandler;
//   private Context context;

//   @Override
//   public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
//     context = binding.getApplicationContext();

//     methodChannel = new MethodChannel(binding.getBinaryMessenger(), METHOD_CHANNEL);
//     methodChannel.setMethodCallHandler(this);

//     eventChannel = new EventChannel(binding.getBinaryMessenger(), EVENT_CHANNEL);
//     speedHandler = new NetworkSpeedHandler(context);
//     eventChannel.setStreamHandler(speedHandler);
//   }

//   @Override
//   public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
//     switch (call.method) {
//       case "startMonitoring":
//         boolean monitorTotalTraffic = call.argument("monitorTotalTraffic");
//         int updateIntervalSeconds = call.argument("updateIntervalSeconds");
//         speedHandler.startMonitoring(monitorTotalTraffic, updateIntervalSeconds);
//         result.success(true);
//         break;

//       case "stopMonitoring":
//         speedHandler.stopMonitoring();
//         result.success(true);
//         break;

//       case "isMonitoring":
//         result.success(speedHandler.isMonitoring());
//         break;

//       case "getMonitoringMode":
//         result.success(speedHandler.getMonitoringMode());
//         break;

//       case "getUpdateInterval":
//         result.success(speedHandler.getUpdateInterval());
//         break;

//       case "setUpdateInterval":
//         int seconds = call.arguments();
//         speedHandler.setUpdateInterval(seconds);
//         result.success(true);
//         break;

//       case "dispose":
//         speedHandler.dispose();
//         result.success(true);
//         break;

//       default:
//         result.notImplemented();
//         break;
//     }
//   }

//   @Override
//   public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
//     if (methodChannel != null) {
//       methodChannel.setMethodCallHandler(null);
//       methodChannel = null;
//     }
//     if (eventChannel != null) {
//       eventChannel.setStreamHandler(null);
//       eventChannel = null;
//     }
//     if (speedHandler != null) {
//       speedHandler.dispose();
//       speedHandler = null;
//     }
//     context = null;
//   }
// }

File: ios/Classes/NetworkSpeedPlusPlugin.swift
import Flutter
import UIKit

import Flutter
import UIKit
import Network

public class NetworkSpeedPlusPlugin: NSObject, FlutterPlugin {
    private static let METHOD_CHANNEL = "com.vinay.network_speed_plus/methods"
    private static let EVENT_CHANNEL = "com.vinay.network_speed_plus/events"

    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var speedHandler: NetworkSpeedHandler?
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: METHOD_CHANNEL,
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: EVENT_CHANNEL,
            binaryMessenger: registrar.messenger()
        )
        let instance = NetworkSpeedPlusPlugin()
        instance.methodChannel = methodChannel
        instance.eventChannel = eventChannel
        instance.speedHandler = NetworkSpeedHandler()

        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance.speedHandler)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startMonitoring":
            guard let args = call.arguments as? [String: Any],
                  let monitorTotal = args["monitorTotalTraffic"] as? Bool,
                  let interval = args["updateIntervalSeconds"] as? Int else {
                result(FlutterError(code: "INVALID_ARGS",
                                   message: "Invalid arguments",
                                   details: nil))
                return
            }
            speedHandler?.startMonitoring(
                monitorTotalTraffic: monitorTotal,
                updateIntervalSeconds: interval
            )
            result(true)

        case "stopMonitoring":
            speedHandler?.stopMonitoring()
            result(true)

        case "isMonitoring":
            result(speedHandler?.isMonitoring() ?? false)

        case "getMonitoringMode":
            result(speedHandler?.getMonitoringMode() ?? "total")

        case "getUpdateInterval":
            result(speedHandler?.getUpdateInterval() ?? 3)

        case "setUpdateInterval":
            guard let seconds = call.arguments as? Int else {
                result(FlutterError(code: "INVALID_ARGS",
                                   message: "Invalid arguments",
                                   details: nil))
                return
            }
            speedHandler?.setUpdateInterval(seconds)
            result(true)

        case "dispose":
            speedHandler?.dispose()
            result(true)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}


// public class NetworkSpeedPlusPlugin: NSObject, FlutterPlugin {
//   public static func register(with registrar: FlutterPluginRegistrar) {
//     let channel = FlutterMethodChannel(name: "network_speed_plus", binaryMessenger: registrar.messenger())
//     let instance = NetworkSpeedPlusPlugin()
//     registrar.addMethodCallDelegate(instance, channel: channel)
//   }
//
//   public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
//     switch call.method {
//     case "getPlatformVersion":
//       result("iOS " + UIDevice.current.systemVersion)
//     default:
//       result(FlutterMethodNotImplemented)
//     }
//   }
// }
